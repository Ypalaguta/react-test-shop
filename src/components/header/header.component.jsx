import {React, conte, useState} from 'react';
import {connect} from 'react-redux';
import {toggleCart} from '../../redux/cart/cart.actions';
import HeaderIcon from '../header-icon/header-icon.styles';
import {Link} from 'react-router-dom';

import {default as HeaderContainer, HeaderLink, HeaderIconWrap, LinkLike} from './header.styles';
import CartComponent from "../cart-popup/cart.component";
import {createStructuredSelector} from "reselect";
import {selectCart} from "../../redux/cart/cart.selectors";
import {selectUserStatus} from "../../redux/user/user.selectors";
import {userSignOutStart} from "../../redux/user/user.actions";

const countAndGetItemsCount = (cartObject) => {
    let items = 0;
    Object.keys(cartObject).forEach(key => {
        let el = cartObject[key];
        items += el.count
    })
    return items ? `(${items})` : '';
}

export function Header(props) {
    const {cart, userStatus, userSignOutStart} = props;
    return (<HeaderContainer>
            <HeaderIconWrap to='/'>
                <Link to='/'>
                    <HeaderIcon src='/pet-icon.jpg'/>
                </Link>
            </HeaderIconWrap>
            <HeaderLink to='/'>Shop</HeaderLink>
            <HeaderLink to='/contacts'>Contacts</HeaderLink>
            {userStatus ? <LinkLike onClick={userSignOutStart}>Sign Out</LinkLike> :
                <HeaderLink to='/auth'>Sign In</HeaderLink>}
            <LinkLike onClick={props.toggleCart}>
                Cart
                {countAndGetItemsCount(cart.items)}
            </LinkLike>
            <CartComponent/>
        </HeaderContainer>
    );
}

const mapStateToProps = createStructuredSelector({
    cart: selectCart,
    userStatus: selectUserStatus
});

export default connect(mapStateToProps, {toggleCart, userSignOutStart})(Header);