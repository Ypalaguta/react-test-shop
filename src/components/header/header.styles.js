import styles from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';

export default styled.div`
    height:70px;
    margin-bottom:10px;
    display: flex;
    justify-content: flex-end;
`;

export const HeaderIconWrap = styled.div`
    flex: auto;
`;

const linkBase = `
    margin-left:10px;
    margin-top:20px;
    height:30px;
    :hover {
        /*text-decoration: underline;*/
        /*background: linear-gradient(270deg,hsl(0 100% 50% / 0.5) 16.66%,rgb(253 140 0 / 0.5) 16.66%,33.32%,rgb(255 229 0 / 0.5) 33.32%,49.98%,rgb(17 159 11 / 0.5) 49.98%,66.64%,rgb(6 68 179 / 0.5) 66.64%,83.3%,rgb(194 46 220 / 0.5) 83.3%);*/
        font-weight: bold;
    }
`;

export const HeaderLink = styled(Link)`
    ${linkBase}
`;

export const LinkLike = styled.div`
    ${linkBase}
    cursor: pointer;
`;