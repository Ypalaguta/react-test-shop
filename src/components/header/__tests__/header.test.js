import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {Link} from 'react-router-dom'
import {HeaderLink, LinkLike} from '../header.styles'

import {Header} from '../header.component';

describe('header rendering', () => {
    let wrapper;
    // beforeEach(()=>{
    //     wrapper = shallow(<Header {...mockProps}/>)
    // })
    const mockProps = {
        cart: {
            items: {}
        },
        userStatus: true,
        toggleCart: jest.fn(),
    }

    test('header structure test', () => {
        wrapper = shallow(<Header {...mockProps}/>)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    test('cart 5 items', () => {
        const testProps = {
            ...mockProps,
            cart: {
                items: {
                    overall1: {
                        category: "overalls",
                        count: 4,
                        id: "overall1",
                        name: "overall 1",
                        pic: "/categories-images/overalls/overall-1.webp",
                        price: "1.5"
                    },
                    overall2: {
                        category: "overalls",
                        count: 1,
                        id: "overall1",
                        name: "overall 1",
                        pic: "/categories-images/overalls/overall-1.webp",
                        price: "1.5"
                    },
                }
            },
        }
        wrapper = shallow(<Header {...testProps}/>)
        expect(wrapper.find(LinkLike).at(1).text()).toEqual('Cart(5)');
        // expect(toJson(wrapper)).toMatchSnapshot()
    })

    test('cart 0 items', () => {
        wrapper = shallow(<Header {...mockProps}/>)
        expect(wrapper.find(LinkLike).at(1).text()).toEqual('Cart')
    })

    test('not signed user', () => {
        wrapper = shallow(<Header {...mockProps} userStatus={false} />)
        expect(wrapper.find(HeaderLink).at(2).text()).toEqual('Sign In')
    })

    test('signed user', () => {
        wrapper = shallow(<Header {...mockProps} userStatus={true} />)
        expect(wrapper.find(LinkLike).at(0).text()).toEqual('Sign Out')
    })
})