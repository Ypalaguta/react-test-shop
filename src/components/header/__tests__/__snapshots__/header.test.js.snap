// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header rendering header structure test 1`] = `
<styled.div>
  <styled.div
    to="/"
  >
    <Link
      to="/"
    >
      <styled.img
        src="/pet-icon.jpg"
      />
    </Link>
  </styled.div>
  <Styled(Link)
    to="/"
  >
    Shop
  </Styled(Link)>
  <Styled(Link)
    to="/contacts"
  >
    Contacts
  </Styled(Link)>
  <styled.div>
    Sign Out
  </styled.div>
  <styled.div
    onClick={[MockFunction]}
  >
    Cart
  </styled.div>
  <Connect(CartComponent) />
</styled.div>
`;
